File: backend/src/config/redis.ts
import { Redis } from 'ioredis';
import { env } from './env.js';

export const redis = new Redis(env.REDIS_URL, {

  maxRetriesPerRequest: null,
  enableReadyCheck: true,
  retryStrategy(times: number): number | null {
    if (times > 10) {
      console.error('[Redis] Max retries reached, giving up');
      return null;
    }
    const delay = Math.min(times * 200, 5000);
    console.log(`[Redis] Retrying connection in ${delay}ms (attempt ${times})`);
    return delay;
  },
});

redis.on('connect', () => {
  console.log('[Redis] Connected successfully');
});

redis.on('error', (err: Error) => {
  console.error('[Redis] Connection error:', err.message);
});

redis.on('close', () => {
  console.log('[Redis] Connection closed');
});

export async function testRedisConnection(): Promise<boolean> {
  try {
    await redis.ping();
    console.log('[Redis] PING successful');
    return true;
  } catch (error) {
    console.error('[Redis] PING failed:', (error as Error).message);
    return false;
  }
}

File: backend/app/config/redis.py
import logging
import fakeredis.aioredis as fakeredis_aio
import redis.asyncio as aioredis
from typing import Optional, Union
from app.config.env import env

# Global Redis client
redis: Optional[Union[aioredis.Redis, fakeredis_aio.FakeRedis]] = None
_using_fake = False

async def init_redis() -> None:
    global redis, _using_fake
    
    # First, try connecting to a real Redis server
    try:
        real_redis = aioredis.from_url(
            env.REDIS_URL,
            decode_responses=True,
            socket_timeout=3.0,
            socket_connect_timeout=3.0,
            retry_on_timeout=False
        )
        await real_redis.ping()
        redis = real_redis
        _using_fake = False
        logging.info("[Redis] Connected to real Redis server successfully")
    except Exception as e:
        # Real Redis not available — fall back to fakeredis
        logging.warning(f"[Redis] Real Redis not available ({e}), using in-memory fakeredis")
        redis = fakeredis_aio.FakeRedis(decode_responses=True)
        _using_fake = True
        logging.info("[Redis] In-memory fakeredis initialized successfully")

async def close_redis() -> None:
    global redis, _using_fake
    if redis:
        try:
            await redis.close()
        except Exception:
            pass
        redis = None
        logging.info("[Redis] Connection closed")

async def test_redis_connection() -> bool:
    global redis
    if not redis:
        try:
            await init_redis()
        except Exception:
            return False
            
    try:
        await redis.ping()
        logging.info("[Redis] PING successful")
        return True
    except Exception as e:
        logging.error(f"[Redis] PING failed: {e}")
        return False
